Add "Sliding Range Join" execution plan

nicktobey opened the pull request:
This is the draft implementation of the "Sliding Range Join" execution plan. This allows for more performant joins when the join condition checks that the column on one table is within a range specified by two columns on the other table.

TODO: Elaborate in this description before making this PR not a draft.

max-hoffman commented on sql/rowexec/rel.go:
```diff
@@ -732,3 +732,17 @@ func (b *BaseBuilder) buildResolvedTable(ctx *sql.Context, n *plan.ResolvedTable
 	return sql.RowsToRowIter(sql.Row{int64(n.Count())}), nil
 }
+
+func (b *BaseBuilder) buildSlidingRange(ctx *sql.Context, n *plan.SlidingRange, row sql.Row) (sql.RowIter, error) {
```
TLDR we want this to construct and return a `RowIter` for the range join that is 1) primed to do work, but hasn't done anything yet, and 2) is ideally a `rowexec` object whose mutations won't impact the original analysis plan objects.
We separate analyzer objects from execution objects. At the end of analysis we have a tree of plan nodes, and that representation is converted into an execution template by the `rowexec` builder. We can make several instances of a single plan execution from the same plan. The template for execution is a volcano row iter, a state machine that will start generating rows when we call `Next()`. The `RowIter` interface, and any implementations (like `filterIter`, `projectIter`, etc) are good examples of how we compartmentalize execution data structures. You could imagine us implementing a `colexec` package in the future, which converts the same analyzer plan to an equivalent but internally different state machine. More practically, this separation makes it easier for me to rewrite the analysis format of the plan, without having to worry about the blast radius reaching the execution format.

nicktobey replied:
I see. So I should not be storing execution state in the `plan.SlidingRange` object, because it's just a template for generating plan executions.

This could be tricky, because we need to maintain state (the priority queue of live ranges) in between each time we load and iterate over the secondary table. We'd also like the join iterator to be agnostic about the execution plan of its secondary child.

Do we have other nodes that have this property? Hash joins do some form of caching, right?

max-hoffman replied:
The trade-off here is maybe (i) if we do not piggy-back off of the default iterator, we have to do more work to add the default null, left join, scope management features to a new `rangeJoinIter` (the same way we do with `mergeJoinIter`, `lateralJoinIter`), or (ii) we will have periodic cache management, isolation, and memory problems with range joins. I personally prefer (i), mostly because (1) our hash join problems have been tricky in the past, (2) I want to rewrite hash join to not use `CachedResults` for performance reasons anyway, and (3) from an architectural perspective I'd rather have execution isolation/speed with new code duplication problems than have analysis and execution interdependent but with speed/cache management problems.

nicktobey replied:
Yeah, I can totally see how stateful / primary-row-dependent secondary iters causes caching bugs that are really hard to debug. I'll create a new `rangeJoinIter`.

max-hoffman commented on sql/memo/memo.go:
```diff
@@ -305,6 +307,20 @@ func (m *Memo) memoizeBindvar(e *expression.BindVar) *ExprGroup {
 	return grp
 }
 
+func (m *Memo) memoizeBetween(e *expression.Between) *ExprGroup {
```
i think we might normalize `Between` into the two halves, might have to check for the Lt/Gt combos

nicktobey replied:
I don't think we do, because in my tests, we are indeed matching on Between expressions and generating the SlidingRangeJoin plan.

However, we probably want to check for the Lt/Gt combos anyway, because Between is always inclusive on both ends, and the client query isn't. This was just faster to initially only support Between.

nicktobey replied:
We now support parsing both Between and LT/GT pairs, see the new PR for details.

max-hoffman commented on sql/memo/exec_builder.go:
```diff
@@ -133,6 +133,53 @@ func (b *ExecBuilder) buildLookupJoin(j *LookupJoin, input sql.Schema, children
+	leftSch := input[:len(input)-len(sr.Parent.Right.RelProps.OutputCols())]
+	rightSch := input[len(sr.Parent.Left.RelProps.OutputCols()):]
+
+	switch n := children[0].(type) {
```
i think the children for this join might be limited to `ResolvedTable` and `TableAlias`. Any PROJECT node can only be pushed into the table. LIMIT and DISTINCT has to wrap the relational expressions (higher in tree). Filter is possible. I might be wrong but I expected the children to be two `IndexedTableAccess` nodes, similar to merge join. It seems like you've noticed this, but the schemas that lookup expressions, join filters, and table filters see will be subtlety different.

nicktobey replied:
This is the same children types that `plan.Lookup` checks for.

I refactored slightly, see the new PR for details.

max-hoffman commented on sql/analyzer/indexed_joins.go:
```diff
@@ -645,13 +649,76 @@ func addHashJoins(m *memo.Memo) error {
+				satisfiesScalarRefs(f.Max.Scalar, join.Right)) {
+				return nil
+			}
+			// TODO: Is this safe? If the expression references multiple columns, does this reference one
```
selection conditions:
- we have one rel with a LessThan/GreaterThan filter combo that sandwiches the same attribute on the other rel
- there are a variety of exclusion conditions for complex expressions that would interfere with range join (ex: `x + 1 <= max AND x > min + 7` is OK because expressions are monotonic increasing (`isWeaklyMonotonic`))
- there are a variety of child node exclusion conditions for each relation (I think tableAlias, resolvedTable, Filter are the 3 OK)
- we have a `sortedIndexForTableCol` for finding appropriate sorted indexes

The most novel steps are the first two, because we don't have any other joins that try to stitch together filters into a Between (which I think have been normalized in a prior analyzer step). Exclusion conditions can be empty to start, we just want to get TableAlias and ResolvedTable into the execution phase for perf testing.

Assembling, we have 1) min expr, 2) max expr, 3) target expr, 4) range table + index (IndexScan), 5) target table + index (IndexScan). Building will fix indexes for 1,2,3, and then (4) and (5) will both be converted into `IndexedTableAccess` for a range spanning NULL->infinity. For mergeJoin I was lazy and put all of the various filter expressions into the join condition in a specific order that I could predictably unpack on the other side.

max-hoffman replied:
Handling sided-ness and left joins will be another set of edge cases

nicktobey replied:
This should be handled in the new PR now.

max-hoffman reviewed (commented):
Excited to see the quick progress! I tried adding some directional comments. For the implementation guts obviously trust yourself if you run into something contrary to my what I said here, managing the selection criteria and inclusion/exclusion conditions is tricky.

max-hoffman commented:
@nicktobey this ain't ready

nicktobey commented:
That wasn't supposed to happen.

I tried to push changes to my branch but there were automated changes already committed to the branch. So I rebased on top of those changes and then pushed.

Except apparently, at some point during that process, I checked out main in order to confirm the behavior at main.
And then I forgot to check out my branch again.

So I ended up rebasing my remote branch onto **main** and then pushing.

Apparently, GitHub interprets this as merging the PR via git.

I fixed the history before anyone noticed (shhh...) But I don't think I can un-merge this PR.

Is it okay if we continue to use this PR for comments/feedback and I create a new PR once they're resolved?

max-hoffman commented:
Sure, whichever is easiest for you.
